Source/TaskDelegate.swift

```diff
@@ -38,19 +38,30 @@ open class TaskDelegate: NSObject {
 
     /// The error generated throughout the lifecyle of the task.
     public var error: Error?
-
-    var task: URLSessionTask? {
+    
+    private var _task: URLSessionTask? {
         didSet { reset() }
     }
-
+    var task: URLSessionTask? {
+        set {
+            taskLock.lock(); defer { taskLock.unlock() }
+            _task = newValue
+        }
+        get {
+            taskLock.lock(); defer { taskLock.unlock() }
+            return _task
+        }
+    }
+    
+    private let taskLock = NSLock()
     var initialResponseTime: CFAbsoluteTime?
     var credential: URLCredential?
     var metrics: AnyObject? // URLSessionTaskMetrics
```

Comment:
Could we pull the `private` properties down underneath the `internal` properties?

```swift
public var error: Error?

var task: URLSessionTask? {
    set {
        taskLock.lock(); defer { taskLock.unlock() }
        _task = newValue
    }
    get {
        taskLock.lock(); defer { taskLock.unlock() }
        return _task
    }
}

var initialResponseTime: CFAbsoluteTime?
var credential: URLCredential?
var metrics: AnyObject? // URLSessionTaskMetrics

private var _task: URLSessionTask? {
    didSet { reset() }
}

private let taskLock = NSLock()

// MARK: Lifecycle
```